tutor-discovery codebase Python 3.12 upgrade

- [x] Ensure tutor-discovery codebase is compatible with Python 3.12
- [x] Update Python to 3.12 in Dockerfile for nightly and ensure the image builds as expected. The changes would be merged to nightly. his task would rely on upgrade of course-discovery to Python 3.12

See https://github.com/overhangio/tutor/issues/1004 for the context.

I attempted building Discovery dockerfile on Python 3.12 on Nightly branch. 
Faced this error mentioned below
```
6.867   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
6.867   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
6.867   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
6.867   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
6.867   File "<frozen importlib._bootstrap_external>", line 995, in exec_module
6.867   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
6.867   File "/openedx/discovery/course_discovery/apps/catalogs/models.py", line 1, in <module>
6.868     from collections import Iterable  # lint-amnesty, pylint: disable=deprecated-class, no-name-in-module
6.868     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6.868 ImportError: cannot import name 'Iterable' from 'collections' (/opt/pyenv/versions/3.12.2/lib/python3.12/collections/__init__.py)
7.026 make: *** [Makefile:20: static] Error 1
------
Dockerfile:81
--------------------
  79 |     # Collect static assets
  80 |     COPY --chown=app:app assets.py ./course_discovery/settings/assets.py
  81 | >>> RUN DJANGO_SETTINGS_MODULE=course_discovery.settings.assets make static
  82 |     
  83 |     # Run production server
```
Above error was fixed by this [PR](https://github.com/openedx/course-discovery/pull/4289) that I created against [discovery](https://github.com/openedx/course-discovery) codebase.

The build was then successful on nightly branch with python 3.12. I was able to access discovery admin panel.
Here's the dockerfile updgrade [PR link](https://github.com/overhangio/tutor-discovery/pull/67) for python 3.12.

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -7,12 +7,15 @@ on:
 jobs:
   tests:
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+         python-version: ['3.8', '3.12']
     steps:
-      - uses: actions/checkout@v2
-      - name: Set up Python 3.8
-        uses: actions/setup-python@v2
+      - uses: actions/checkout@v3
+      - name: Set up Python
+        uses: actions/setup-python@v5
         with:
-          python-version: 3.8
+          python-version:  ${{ matrix.python-version }}
       - name: Upgrade pip
         run: python -m pip install --upgrade pip setuptools
       - name: Install dependencies

--- a/changelog.d/20240313_150946_danyal.faheem_upgrade_to_python_3_12.md
+++ b/changelog.d/20240313_150946_danyal.faheem_upgrade_to_python_3_12.md
@@ -1,0 +1,1 @@
+- [Feature] Add Python 3.12 CI checks (by @Danyal-Faheem)

